sdk/nodejs/resource.ts

```diff
@@ -13,13 +13,34 @@
 // limitations under the License.
 
 import { ResourceError, RunError } from "./errors";
-import { Input, Inputs, Output } from "./output";
+import { all, Input, Inputs, interpolate, Output, output } from "./output";
 import { readResource, registerResource, registerResourceOutputs } from "./runtime/resource";
+import { getProject, getStack } from "./runtime/settings";
 import * as utils from "./utils";
 
 export type ID = string;  // a provider-assigned ID.
 export type URN = string; // an automatically generated logical URN, used to stably identify resources.
 
+/**
+ * createUrn computes a URN from the combination of a resource name, resource type, optional parent,
+ * optional project and optional stack.
+ */
+export function createUrn(name: Input<string>, type: Input<string>, parent?: Resource | Input<URN>, project?: string, stack?: string): Output<string> {
+    let parentPrefix: Output<string>;
+    if (parent) {
+        let parentUrn: Output<string>;
+        if (Resource.isInstance(parent)) {
+            parentUrn = parent.urn;
+        } else {
+            parentUrn = output(parent);
+        }
+        parentPrefix = parentUrn.apply(parentUrnString => parentUrnString.substring(0, parentUrnString.lastIndexOf("::")) + "$");
+    } else {
+        parentPrefix = output(`urn:pulumi:${stack || getStack()}::${project || getProject()}::`);
+    }
```

Comment:
i keep looking at both paths above and something seems totally busted to me.  Through the 'true' path you grab hte parent urn, find the last `::` and then substr *up to* that point.  So the true-path should not end with `::` afaict. But through the lower path, you form the prefix always has the trailing `::`.  I think it would be very good to have some examples here to demonstrate why this is felt to be correct.

Reply:
It's subtle - and this definitely needs tests! - but I believe it is correct. The `::` in the first branch is the final one (before the name component).  The `::` in the second branch is the next-to-final one (before the type).  The final `::` is always added in the line below.

Definitely happy to rewrite this to a more complete semantic tear-down and recreate of the URN - but was trying to keep this pretty raw since ideally we replace this with a call back into the engine at some point so we can centralize URN handling.